osquery/events/eventfactory.cpp

```diff
@@ -0,0 +1,478 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/config/config.h>
+#include <osquery/core/flags.h>
+#include <osquery/core/system.h>
+#include <osquery/events/eventfactory.h>
+#include <osquery/events/eventsubscriber.h>
+#include <osquery/events/types.h>
+#include <osquery/logger/logger.h>
+#include <osquery/registry/registry.h>
+#include <osquery/sql/sql.h>
+
+namespace osquery {
+namespace {
+/**
+ * @brief Details for each subscriber as it relates to the schedule.
+ *
+ * This is populated for each configuration update by scanning the schedule.
+ */
+struct SubscriberExpirationDetails {
+ public:
+  /// The max internal is the minimum wait time for expiring subscriber data.
+  size_t max_interval{0};
+
+  /// The number of queries that should run between intervals.
+  size_t query_count{0};
+};
+} // namespace
+
+FLAG(bool, disable_events, false, "Disable osquery publish/subscribe system");
+
+// There's no reason for the event factory to keep multiple instances.
+EventFactory& EventFactory::getInstance() {
+  static EventFactory ef;
+  return ef;
+}
+
+Status EventFactory::registerEventSubscriber(const PluginRef& sub) {
+  // Try to downcast the plugin to an event subscriber.
+  EventSubscriberRef specialized_sub;
+  try {
+    auto base_sub = std::dynamic_pointer_cast<EventSubscriberPlugin>(sub);
+    specialized_sub = std::static_pointer_cast<BaseEventSubscriber>(base_sub);
+  } catch (const std::bad_cast& /* e */) {
+    return Status(1, "Incorrect plugin");
+  }
+
+  if (specialized_sub == nullptr || specialized_sub.get() == nullptr) {
+    return Status(1, "Invalid subscriber");
+  }
+
+  // The config may use an "events" key to explicitly enabled or disable
+  // event subscribers. See EventSubscriber::disable.
+  auto name = specialized_sub->getName();
+  if (name.empty()) {
+    // This subscriber did not override its name.
+    return Status(1, "Subscribers must have set a name");
+  }
+
+  auto plugin = Config::get().getParser("events");
+  if (plugin != nullptr && plugin.get() != nullptr) {
+    const auto& data = plugin->getData().doc();
+    // First perform explicit enabling.
+    if (data["events"].HasMember("enable_subscribers")) {
+      for (const auto& item : data["events"]["enable_subscribers"].GetArray()) {
```

Comment:
You may just be moving code, but while you are here this could use a check before `IsArray()` before we call `GetArray()`.

Reply:
see https://github.com/osquery/osquery/pull/6610#discussion_r540330119